File: designPatterns/Behavioral_patterns/Memento/javascript/ConcreteMemento.js
const Memento = require("./Memento");

/**
 * The Concrete Memento contains the infrastructure for storing the Originator's
 * state.
 */
class ConcreteMemento extends Memento {
  constructor(state) {
    super();
    this.state = state;
    this.date = new Date().toISOString().slice(0, 19).replace("T", " ");
  }

  /**
   * The Originator uses this method when restoring its state.
   */
  getState() {
    return this.state;
  }

  /**
   * The rest of the methods are used by the Caretaker to display metadata.
   */
  getName() {
    return `${this.date} / (${this.state.substr(0, 9)}...)`;
  }

  getDate() {
    return this.date;
  }
}

module.exports = ConcreteMemento;

File: designPatterns/Behavioral_patterns/Memento/python/ConcreteMemento.py
from datetime import datetime
from Memento import Memento


class ConcreteMemento(Memento):
    def __init__(self, state):
        self._state = state
        self._date = str(datetime.now())[:19]

    def get_state(self):
        """
        The Originator uses this method when restoring its state.
        """
        return self._state

    def get_name(self):
        """
        The rest of the methods are used by the Caretaker to display metadata.
        """

        return f"{self._date} / ({self._state[0:9]}...)"

    def get_date(self):
        return self._date
